Feature: sidecar for ContainerOp

eterna2 opened the pull request:
## Motivation
I am using kubeflow pipelines to run simulation experiments, and because of the complexity of the simulation setup, it requires supporting containers which I might want to parameterize. 

## Status
Waiting for Review
- all unit tests are working
- updated to work with latest master HEAD
- TODO?: more comprehensive test on generating complex template with PipelineParams everywhere.

## New or updated Features
This PR adds the following features to `kfp` package's `dsl` and `compiler` modules:
- allows `PipelineParam` for all valid attributes/properties (both string serialized and object instance) in `dsl.ContainerOp`
- allows sidecars for workflow template (via `dsl.ContainerOp.add_sidecar`) with `dsl.Sidecar` class 
- allows parameterization of `container` property for argo workflow template (via `dsl.ContainerOp.container`) 
 (i.e. you can set any valid attributes/properties for a k8s `V1Container` object)

> NOTE . 
> All changes are completely backward compatible. With appropriate pending deprecation warnings.

Example
```py
from kfp import dsl
from kubernetes.client.models import V1EnvVar

@dsl.pipeline(
    name='foo',
    description='hello world')
def foo_pipeline(tag: str, pull_image_policy: str):

    # any attributes can be parameterized (both serialized string or actual PipelineParam)
    op = dsl.ContainerOp(name='foo', 
                                         image='busybox:' % tag,
                                         # pass in sidecars list
                                         sidecars=[dsl.Sidecar('print', 'busybox:latest', command='echo "hello"')],
                                         # pass in k8s container kwargs
                                         container_kwargs={'env': [V1EnvVar('foo', 'bar')]})

    # set `imagePullPolicy` property for `container` with `PipelineParam` 
    op.container.set_pull_image_policy(pull_image_policy)

    # add sidecar with parameterized image tag
    # sidecar follows the argo sidecar swagger spec
    op.add_sidecar(dsl.Sidecar('redis', 'redis:' % tag).set_image_pull_policy('Always'))
```

## TLDR changes
This PR will make the following changes:
- Created `Container` and `Sidecar` classes
 which inherits from `V1Container`
  - updated both classes with methods to update k8s properties

- Updated `ContainerOp`
  - added kwargs: `container_kwargs` and `sidecars`
  - added attr `_container` (and getter `container`) which holds the `Container` instance
  - added `pendingDeprecationWarning` for container properties and methods
  - changed `inputs` to scan through all qualified attributes in `ContainerOp` to get all `PipelineParam`

- Updated `PipelineParam`
  - added `pattern` attr to hold regex pattern extracted from serialized string
  - added method to recursively extract `PipelineParam` from any object

- ContainerOp to template 
  - moved `_op_to_template` to its own module
  - added method to recursively replace `PipelineParam` (and its serialized form) with appropriate `input.parameters.%s`

- Added appropriate unit test for updates

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/879)
<!-- Reviewable:end -->

Ark-kun commented on sdk/python/kfp/compiler/compiler.py:
```diff
@@ -332,6 +332,36 @@ def _build_conventional_artifact(name):
+      def _sidecar_to_template(tsidecar):
+          index, sidecar = tsidecar
+          sidecar_template = {
+            'name': f'{op.name}-{sidecar.name}-{index}',
```
Not sure this format is supported in Python 3.5

eterna2 replied:
replaced `f-string` with `string format` for py 3.5.

eterna2 commented:
@animeshsingh

Asynchronously. The sidecars are services rather than jobs (i.e. mq brokers, db, rest services, ...)

I am running a simulation job that is reliant on a few other services. If I provision these services as parallel steps, I would need to add additional scripts to send a termination signal to these services when the simulation job completes.

hongye-sun commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -20,88 +20,21 @@
-  def __init__(self, name: str, image: str, command: str=None, arguments: str=None,
-               file_outputs : Dict[str, str]=None, is_exit_handler=False):
-    """Create a new instance of ContainerOp.
+class ContainerBase(object):
```
This should be the k8s Container spec. How about just call it Container?

hongye-sun commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -249,6 +170,132 @@ def add_env_variable(self, env_variable):
     return self
 
+
+class SideCar(ContainerBase):
```
SideCar basically is just a k8s Container. You don't really need to specialize it to a type. Can you move all the fields in it to the Container base type?

hongye-sun replied:
Actually, sidecar in argo is slightly different from k8s container. You still need to keep the class but move the arguments and command to base class. Could you also add mirrorVolumeMounts field to sidecar, which is the only field which is addition to container spec?

eterna2 replied:
Sure. Is there somewhere I can see the Argo spec for sidecar? I tried looking at the docs, but it is so sparse - just the example.

Hence, I have no idea the full spec for sidecar.

eterna2 replied:
nvm. I found it. I will try to implement the more common/important ones.

https://github.com/argoproj/argo/blob/master/api/openapi-spec/swagger.json

hongye-sun commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -249,6 +170,132 @@ def add_env_variable(self, env_variable):
+    self.sidecars = []
+
+    matches = []
+    for arg in (command or []) + (arguments or []):
```
We actually have a plan to extract the pipeline param from the whole body of the ContainerOp instead of just command and arguments. Argo supports placeholders for the whole template spec. If you don't want to expand the scope of the PR, you might create a utility method to extract param from a container's command and arguments and apply it to ContainerOp (for the main container) and sidecar containers here.

eterna2 replied:
@hongye-sun 
Hey, I have made a very big change to the PR to cater to this feature. 

Essentially, it is a recursive scan of `ContainerOp` as well as any k8s object and cached inside `inputs`.

I also replaced the `container` prop in argo workflow template with a wrapped k8s `V1Container`, so that any valid k8s prop can be set for `ContainerOp`.

hongye-sun replied:
@eterna2, awesome, really appreciate your contribution here.

hongye-sun commented on sdk/python/kfp/compiler/compiler.py:
```diff
@@ -332,6 +332,36 @@ def _build_conventional_artifact(name):
       template['retryStrategy'] = {'limit': op.num_retries}
 
+    if op.sidecars:
+      def _sidecar_to_template(tsidecar):
```
Can I ask if you can extract the code to be a common utility to convert a Container to k8s container json spec and reuse the method for both main container and sidecars? Thanks.

cliveseldon commented:
Is it worth adding an example or test to illustrate how a sidecar can be added?

eterna2 commented:
@cliveseldon 
Sure. I will add an @example as part of the pydoc comment for ContainerOp.

animeshsingh commented on sdk/python/kfp/dsl/__init__.py:
```diff
@@ -15,6 +15,6 @@
 from ._pipeline_param import PipelineParam
 from ._pipeline import Pipeline, pipeline, get_pipeline_conf
-from ._container_op import ContainerOp
+from ._container_op import ContainerOp, SideCar
```
Given Argo uses 'Sidecar' may make sense to use small 'c' to show compatibility?

animeshsingh commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -249,6 +170,132 @@ def add_env_variable(self, env_variable):
+
+
+class ContainerOp(ContainerBase):
+  """Represents an op implemented by a docker container image."""
```
Given docker is being phased out, may make sense to just call these 'container image'

Ark-kun replied:
> Given docker is being phased out, may make sense to just call these 'container image'

Can you please provide some link documenting that?

animeshsingh replied:
well @Ark-kun - its phased out from IBM`s official Kubernetes service 
https://www.ibm.com/blogs/bluemix/2019/01/ibm-cloud-kubernetes-service-supports-containerd/

Other providers are also discussing moving away - btw GCP lauched this in beta in Nov as well
https://cloud.google.com/blog/products/containers-kubernetes/containerd-available-for-beta-testing-in-google-kubernetes-engine

texasmichelle commented:
👍 to adding sidecar support. Thank you everyone for working on this!

gaoning777 commented:
/hold
Thanks for the contribution. Really great work.
Put a hold here so that I can review before the PR merge to make sure that it does not break any ongoing DSL features.
I will review asap.

eterna2 commented on sdk/python/tests/compiler/testdata/with_sidecars_and_pipelineparams.yaml:
```diff
@@ -0,0 +1,227 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+apiVersion: argoproj.io/v1alpha1
+spec:
```
Nope. Not working. I will try with simpler examples first, before getting it to work with a more complex one.

eterna2 replied:
I removed this, and have a simpler argo workflow template
- 1 for sidecar
- 1 for pipeline params

eterna2 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
+                 command: StringOrStringList = None,
+                 arguments: StringOrStringList = None,
+                 file_outputs: Dict[str, str] = None,
+                 is_exit_handler=False):
```
I probably should call it `container_kwargs`, instead of `**kwargs`? So that we don't mess up with the workflow kwargs?

hongye-sun replied:
sg

eterna2 commented:
@gaoning777

Sure. But I think I might push a few more commits cuz it seems some of the auto-merge and conflicts were not resolved correctly.

I found a few bugs I need to fix. 

I forgot to proxy command, arguments, and image too. Which broke some of the new tests.

I was working on a older commit. But I have gonna all of them to work now. 

But I still need to fix the PipelineParam generations for some edge cases.

eterna2 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
-    self.resource_requests[resource_name] = value
-    return self
+      # creates a operation
+      op = ContainerOp(name='bash-ops', i
```
fixed

eterna2 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
+        # `container` prop in `io.argoproj.workflow.v1alpha1.Template`
+        self._container = Container(
+            image=image, args=arguments, command=command)
+        # for chaining, and returning back to `ContainerOp` when updating `Container`
```
`op.apply` means `op` is already created. U can't do chaining like:
```py

op = (ContainerOp(...)
            .container
                .set_pull_image_policy('Always')
                .add_env_variables(...)
            .parent)
```

But I defer to ur decision. I can remove this.

hongye-sun replied:
Not sure if I get your point right. Isn't it the same code as:
```
op = ContainerOp(...).apply(lambda op: (
  op.container
    .set_pull_image_policy('Always')
    .add_env_variables(...)))
```
It doesn't look too bad with the apply pattern to me, compared with the parent..

eterna2 replied:
ops. Yup. U are right. I mis-read the func. I will remove it.

eterna2 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
+        def _proxy(proxy_attr):
+            """Decorator func to proxy to ContainerOp.container"""
+
+            def _decorated(*args, **kwargs):
```
done

eterna2 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
-  def add_volume_mount(self, volume_mount):
-    """Add volume to the container
+        # creates a `ContainerOp` and adds a redis `Sidecar`
+        op = (ContainerOp(name='foo-op', image='busybox:latest')
```
done: added kwarg `sidecars`

eterna2 commented:
@hongye-sun 

I have resolved all the conflicts with the latest master HEAD, and added appropriate unit tests. 

Everything seems to be working.

hongye-sun commented on sdk/python/kfp/compiler/compiler.py:
```diff
@@ -549,7 +432,7 @@ def _compile(self, pipeline_func):
       sanitized_name = K8sHelper.sanitize_k8s_name(op.name)
       op.name = sanitized_name
-      for param in op.inputs + op.argument_inputs:
+      for param in op.inputs: # + op.argument_inputs:
```
Do you still need this?

eterna2 replied:
No. I think I missed this when resolving the merge conflict.

hongye-sun commented on sdk/python/kfp/dsl/__init__.py:
```diff
@@ -13,9 +13,9 @@
 
 
-from ._pipeline_param import PipelineParam
+from ._pipeline_param import PipelineParam, _match_serialized_pipelineparam
```
Why is it needed to be exposed? If it's really needed, can it be named as a public method?

eterna2 replied:
I need it in `compiler` to extract and replace the serialized pipelineparam with `input.parameters.%s`. Did not want to have 2 sets of duplicated regex in 2 diff modules.

**Proposed solution:**  
We can probably a `StringWithPipelineParams` class (which holds a string with 1 or more serialized PipelineParams) instead of having a mix of strings and `PipelineParam`.

And add `to_argo_variable` method to both classes.

hongye-sun commented:
/retest

It seems that the e2e tests are stuck on the PR.

hongye-sun commented:
/ok-to-test

vicaire commented:
@eterna2, thanks a lot for contributing this. This is great. 

@texasmichelle needs this feature for an important demo that will take place soon and would like to rely on your change instead of implementing a workaround. Do you have an approximate ETA for resolving the remaining conflicts?  Thanks!

texasmichelle commented:
@eterna2 thank you for this PR! Would love to highlight this feature at Next if we can get it in ASAP.

eterna2 commented:
@hongye-sun 
hey I have resync my code with latest HEAD (with ning's commit). Looked thru the changes, don't think have any significant conflicts. Fixed the unit test to call the updated method. 

- made `match_serialized_pipelineparam` public (u can see my proposed solution to clean up the mix between serialized pipeline params and pipelineparam, if u are ok, I can fork a branch and fix that).

- removed `parent` from `ContainerOp.container`

- fixed some minor typos in the comments, and clean some unused imports and stuff I did not clean up properly last time.

Unit tests *should* work (I am on another rig where I did not setup the unit test env locally).

hongye-sun commented:
@eterna2 

The PR lgtm. I am waiting for all the unit and e2e tests to pass to approve it. Thanks.

hongye-sun commented:
/lgtm
/approve

Travis has an outage right now. Since the e2e is passing, I will give the approve label and feel free to merge after travis build is passed.

hongye-sun commented:
/test kubeflow-pipeline-sample-test

gaoning777 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,940 @@
+    def _validate_cpu_string(self, cpu_string):
+        "Validate a given string is valid for cpu request or limit."
+
+        if isinstance(cpu_string, _pipeline_param.PipelineParam):
```
Could we extract this block of code as a separate internal function because it has been used by many functions here?

gaoning777 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,940 @@
+    return value if isinstance(value, list) else [value]
+
+
+def create_and_append(current_list: Union[List[T], None], item: T) -> List[T]:
```
If this is internally used in this module, could we rename it such that it starts with an underscore?

gaoning777 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,940 @@
+        self.image_pull_policy = image_pull_policy
+        return self
+
+    def add_port(self, container_port):
```
Maybe add a TODO to support pipelineparam for the other configurations besides CPU, memory ,etc.

eterna2 replied:
It is already supported - in the `compiler._op_to_template` there is a recursive method to replace `PipelineParam` and its serialized form into argo variables.

gaoning777 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,940 @@
 
-    if re.match(r'^[0-9]+m$', cpu_string) is not None:
-      return
+# util functions
```
Let's move the general utility functions into a separate module.

gaoning777 replied:
This is more container_op related but I think it would be helpful to define a decorator that helps for the future deprecation warnings.

gaoning777 commented on sdk/python/kfp/dsl/_container_op.py:
```diff
@@ -12,309 +12,843 @@
+        #   op._container.set_image_policy('Always').parent == op   # True
+        setattr(self._container, "parent", self)
+
+        # NOTE for backward compatibility (remove in future?)
```
I love the deprecation warnings while it maintains backward compatibility.
It helps to warn the users that these functions will be deprecated.

hongye-sun commented:
/test kubeflow-pipeline-sample-test

hongye-sun commented:
/test kubeflow-pipeline-sample-test

hongye-sun commented:
/hold cancel

hongye-sun commented:
@eterna2 

Thanks a lot for contributing to this PR. It helps to unlock many our use cases (not only just sidecar) for k8s resource orchestrations.

vicaire commented:
+1. Thank you @eterna2
